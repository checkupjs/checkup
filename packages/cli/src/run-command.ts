import * as Config from '@oclif/config';
import * as stdout from 'stdout-monkey';

const castArray = <T>(input?: T | T[]): T[] => {
  if (input === undefined) return [];
  return Array.isArray(input) ? input : [input];
};

let root = require.resolve('.');

export async function runCommand(args: string[] | string, opts: loadConfig.Options = {}) {
  let output: string;
  let patch;

  if (!opts.testing) {
    patch = stdout((str: string) => {
      output += str;
    });
  }

  let config = await Config.load(opts.root || root);

  args = castArray(args);

  const [id, ...extra] = args;
  await config.runHook('init', { id, argv: extra });
  await config.runCommand(id, extra);

  if (!opts.testing) {
    patch.restore();

    return { stdout: output! };
  }

  return;
}

export namespace loadConfig {
  export let root: string;
  export interface Options {
    root?: string;
    reset?: boolean;
    testing?: boolean;
  }
}
