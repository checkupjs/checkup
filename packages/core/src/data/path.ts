export function trimCwd(path: string, cwd: string) {
  return path.replace(`${cwd}/`, '');
}

export function trimAllCwd(paths: string[], cwd: string) {
  return paths.map((path) => trimCwd(path, cwd));
}
