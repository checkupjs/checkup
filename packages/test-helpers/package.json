{
  "name": "@checkup/test-helpers",
  "description": "Checkup's test helper library",
  "version": "0.0.10",
  "author": "Steve Calvert <steve.calvert@gmail.com>",
  "bugs": "https://github.com/checkupjs/checkup/issues",
  "dependencies": {
    "@checkup/core": "0.0.10",
    "fixturify-project": "^2.1.0",
    "handlebars": "^4.7.6",
    "stdout-stderr": "^0.1.13",
    "tmp": "^0.2.0"
  },
  "engines": {
    "node": ">= 12.*"
  },
  "files": [
    "/lib",
    "/static"
  ],
  "homepage": "https://github.com/checkupjs/checkup",
  "license": "MIT",
  "main": "lib/index.js",
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/checkupjs/checkup",
  "scripts": {
    "build": "tsc --build",
    "test": "jest --passWithNoTests --no-cache"
  },
  "types": "lib/index.d.ts"
}
