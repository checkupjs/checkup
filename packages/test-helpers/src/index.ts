export { default as Project } from './ember-cli-fixturify-project';
export { stdout } from './stdout';
export { default as CheckupProject } from './checkup-fixturify-project';
export { default as Plugin } from './plugin';
