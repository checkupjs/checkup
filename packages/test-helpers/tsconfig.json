{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "types": ["chai", "jest", "node"],
    "esModuleInterop": true
  },
  "references": [
    {
      "path": "../core"
    }
  ],
  "include": ["src/**/*"]
}
