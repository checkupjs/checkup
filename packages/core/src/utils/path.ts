import { isAbsolute, join, relative } from 'path';

export function toAbsolute(baseDir: string, filePath: string) {
  let absolutePath = filePath;

  if (!isAbsolute(filePath)) {
    absolutePath = join(baseDir, filePath);
  }

  return absolutePath;
}

export function toRelative(baseDir: string, filePath: string) {
  let relativePath = filePath;

  if (isAbsolute(relativePath)) {
    relativePath = relative(baseDir, filePath);
  }

  return relativePath;
}
