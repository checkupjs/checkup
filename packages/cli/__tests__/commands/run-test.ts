import * as fs from 'fs';
import * as path from 'path';

import { CheckupProject, createTmpDir, stdout } from '@checkup/test-helpers';

import { runCommand } from '../__utils__/run-command';

const TEST_TIMEOUT = 100000;

describe('@checkup/cli', () => {
  describe('normal cli output', () => {
    let project: CheckupProject;

    beforeEach(function () {
      project = new CheckupProject('checkup-project', '0.0.0').addCheckupConfig({
        plugins: [],
        tasks: {},
      });

      project.writeSync();
    });

    afterEach(function () {
      project.dispose();
    });

    it(
      'should output checkup result',
      async () => {
        await runCommand(['run', project.baseDir]);

        expect(stdout()).toMatchSnapshot();
      },
      TEST_TIMEOUT
    );

    it('should output checkup result in JSON', async () => {
      await runCommand(['run', '--reporter', 'json', project.baseDir]);

      expect(stdout()).toMatchSnapshot();
    });

    it(
      'should output an html file in the current directory if the html reporter option is provided',
      async () => {
        await runCommand(['run', '--reporter', 'html', project.baseDir]);

        let outputPath = stdout().trim();

        expect(outputPath).toMatch(
          /^(.*)\/checkup-report-(\d{4})-(\d{2})-(\d{2})-(\d{2})-(\d{2})-(\d{2})\.html/
        );
        expect(fs.existsSync(outputPath)).toEqual(true);

        fs.unlinkSync(outputPath);
      },
      TEST_TIMEOUT
    );

    it(
      'should output an html file in a custom directory if the html reporter and reporterOutputPath options are provided',
      async () => {
        let tmp = createTmpDir();

        await runCommand(['run', '--reporter', 'html', `--reportOutputPath`, tmp, project.baseDir]);

        let outputPath = stdout().trim();

        expect(outputPath).toMatch(
          /^(.*)\/checkup-report-(\d{4})-(\d{2})-(\d{2})-(\d{2})-(\d{2})-(\d{2})\.html/
        );
        expect(fs.existsSync(outputPath)).toEqual(true);

        fs.unlinkSync(outputPath);
      },
      TEST_TIMEOUT
    );

    it('should run a single task if the task option is specified', async () => {
      await runCommand(['run', '--task', 'project', project.baseDir]);

      expect(stdout()).toMatchSnapshot();
    });

    it('should use the config at the config path if provided', async () => {
      const anotherProject = new CheckupProject('another-project').addCheckupConfig({
        plugins: [],
        tasks: {},
      });
      anotherProject.writeSync();
      await runCommand([
        'run',
        '--config',
        path.join(anotherProject.baseDir, '.checkuprc'),
        project.baseDir,
      ]);

      expect(stdout()).toMatchSnapshot();
      anotherProject.dispose();
    });
  });
});
