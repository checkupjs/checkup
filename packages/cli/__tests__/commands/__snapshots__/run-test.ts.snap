// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@checkup/cli normal cli output if excludePaths are provided by both the config and command line, use command line 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (4 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 1 lines of code
■ hbs (1)


checkup v0.0.0
config 0e3da47ee1e0a5e663465af5cc66cf37

"
`;

exports[`@checkup/cli normal cli output if excludePaths are provided by both the config and command line, use command line 2`] = `
"
Checkup report generated for checkup-app v0.0.0 (5 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 2 lines of code
■ js (2)


checkup v0.0.0
config 0e3da47ee1e0a5e663465af5cc66cf37

"
`;

exports[`@checkup/cli normal cli output should output checkup result in JSON 1`] = `
Object {
  "$schema": "https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0-rtm.5.json",
  "properties": Object {
    "actions": Array [],
    "analyzedFiles": Array [
      ".checkuprc",
      ".eslintrc.js",
      "index.hbs",
      "index.js",
      "package-lock.json",
      "package.json",
    ],
    "analyzedFilesCount": 6,
    "cli": Object {
      "args": Object {
        "paths": Array [],
      },
      "config": Object {
        "$schema": "https://raw.githubusercontent.com/checkupjs/checkup/master/packages/core/src/schemas/config-schema.json",
        "excludePaths": Array [],
        "plugins": Array [],
        "tasks": Object {},
      },
      "configHash": "dd17cda1fc2eb2bc6bb5206b41fc1a84",
      "flags": Object {
        "format": "json",
        "outputFile": "",
      },
      "schema": 1,
      "version": "0.0.0",
    },
    "project": Object {
      "name": "checkup-app",
      "repository": Object {
        "activeDays": "0 days",
        "age": "0 days",
        "linesOfCode": Object {
          "total": 3,
          "types": Array [
            Object {
              "extension": "js",
              "total": 2,
            },
            Object {
              "extension": "hbs",
              "total": 1,
            },
          ],
        },
        "totalCommits": 0,
        "totalFiles": 0,
      },
      "version": "0.0.0",
    },
    "timings": Object {},
  },
  "runs": Any<Array>,
  "version": "2.1.0",
}
`;

exports[`@checkup/cli normal cli output should output checkup result in verbose mode 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should output list of available tasks 1`] = `
"
AVAILABLE TASKS

  fake/file-count
  fake/foo

"
`;

exports[`@checkup/cli normal cli output should run a single task if the tasks option is specified with a single task 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 2

File Count Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run a task if its passed in via command line, even if it is turned "off" in config 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 2

File Count Task

■ hi (0)


checkup v0.0.0
config 2910e38984c8092494a42c88b642b9de

"
`;

exports[`@checkup/cli normal cli output should run multiple tasks if the category option is specified with multiple categories 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 2

File Count Task

■ hi (0)

=== Fake 1

Foo Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run multiple tasks if the group option is specified with multiple groups 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 2

File Count Task

■ hi (0)

=== Fake 1

Foo Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run multiple tasks if the tasks option is specified with multiple tasks 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 2

File Count Task

■ hi (0)

=== Fake 1

Foo Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run only one task if the category option is specified 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 1

Foo Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run only one task if the group option is specified 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)

=== Fake 1

Foo Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run the tasks on the globs passed into checkup, if provided, instead of entire app 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (3 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ hbs (2)
■ js (1)

=== Fake 2

File Count Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should run the tasks on the globs passed into checkup, if provided, instead of entire app 2`] = `
"
Checkup report generated for checkup-app v0.0.0 (9 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 6 lines of code
■ js (4)
■ hbs (2)

=== Fake 2

File Count Task

■ hi (0)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should use the config at the config path if provided 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should use the excludePaths provided by the command line 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (3 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

 0 lines of code


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should use the excludePaths provided by the command line 2`] = `
"
Checkup report generated for checkup-app v0.0.0 (5 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 2 lines of code
■ js (2)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;

exports[`@checkup/cli normal cli output should use the excludePaths provided by the config 1`] = `
"
Checkup report generated for checkup-app v0.0.0 (5 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 2 lines of code
■ js (2)


checkup v0.0.0
config 0e3da47ee1e0a5e663465af5cc66cf37

"
`;

exports[`@checkup/cli normal cli output should use the excludePaths provided by the config 2`] = `
"
Checkup report generated for checkup-app v0.0.0 (6 files analyzed)

This project is 0 days old, with 0 days active days, 0 commits and 0 files.

■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3 lines of code
■ js (2)
■ hbs (1)


checkup v0.0.0
config dd17cda1fc2eb2bc6bb5206b41fc1a84

"
`;
